driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorFlux.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.reactivestreams.client.internal;
+
+import com.mongodb.lang.Nullable;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+class BatchCursorFlux<T> implements Publisher<T> {
+
+    private final BatchCursorPublisher<T> batchCursorPublisher;
+    private final Integer originalBatchSize;
+    private final Runnable resetBatchSize;
+    private final AtomicBoolean inProgress = new AtomicBoolean(false);
+    private final AtomicLong demandDelta = new AtomicLong(0);
+    private BatchCursor<T> batchCursor;
```

Comment:
This plain field is modified from
```java
sink.onRequest(l -> {
    ...
    batchCursorPublisher.batchCursor().subscribe(bc -> {
        ...
        batchCursor = bc;
        ...
    })
    ...
})
```
and is read from `BatchCursorFlux.closeCursor` and other places.

I have no knowledge of project Reactor, which means I can't tell whether it guarantees all the necessary happens-before relations for us, so I am pointing this out just in case. @rozza, if you thought about this and it should be fine, then I trust it is.

Reply:
Correct, we set the field when we lazily fetch the cursor.

We explicitly close the batchCursor if the sink is cancelled / disposed of. If calling `batchCursorPublisher.batchCursor().subscribe` produces an error we flag that up to the sink as an error. In that case batchCursor won't have been set.

Reply:
This thread continues here: https://github.com/mongodb/mongo-java-driver/pull/652#discussion_r568185193